docs/contracts/transient-storage.rst

```diff
@@ -0,0 +1,166 @@
+The values cannot be conserved past the end of a transaction.
+
+Transient storage variables have completely independent address space from storage,
+so that the order of transient state variables does not affect the layout of storage
+state variables and vice-versa. They do need distinct names though because all state
+variables share the same namespace.
+It is also important to note that the values in transient storage are packed in the
+same fashion as those in persistent storage.
+See :ref:`Storage Layout <storage-inplace-encoding>` for more information.
+
+Besides that, transient variables can have visibility as well and ``public`` ones will
+have a getter function generated automatically as usual.
+
+Note that such use of ``transient`` as a data location is only allowed for
+:ref:`value type <value-types>` state variable declarations.
+Reference types, such as arrays, mappings and structs, as well as local or parameter
+variables are not supported.
+
+An expected canonical use case for transient storage is cheaper reentrancy locks,
+which can be readily implemented with the opcodes as showcased next.
+
+.. code-block:: solidity
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity ^0.8.27;
+
+    contract Generosity {
+        mapping(address => bool) sentGifts;
+        bool transient locked;
+
+        modifier nonReentrant {
+            require(!locked, "Reentrancy attempt");
+            locked = true;
+            _;
+            // Unlocks the guard, making the pattern composable.
+            // After the function exits, it can be called again, even in the same transaction.
+            locked = false;
+        }
+
+        function claimGift() nonReentrant public {
+            require(address(this).balance >= 1 ether);
+            require(!sentGifts[msg.sender]);
+            (bool success, ) = msg.sender.call{value: 1 ether}("");
+            require(success);
+
+            // In a reentrant function, doing this last would open up the vulnerability
+            sentGifts[msg.sender] = true;
+        }
+    }
+
+Transient storage is private to the contract that owns it, in the same way as persistent storage.
+Only owning contract frames may access their transient storage.
+And when they do, all the frames access the same transient store.
+
+Transient storage is part of the EVM state and is subject to the same mutability enforcements
+as persistent storage. As such, any read access to it is not ``pure`` and writing access is not ``view``.
+
+If the ``TSTORE`` opcode is called within the context of a ``STATICCALL``,
+it will result in an exception instead of performing the modification.
+``TLOAD`` is allowed within the context of a ``STATICCALL``.
+
+When transient storage is used in the context of ``DELEGATECALL`` or ``CALLCODE``,
+then the owning contract of the transient storage is the contract that issued ``DELEGATECALL``
+or ``CALLCODE`` instruction (the caller) as with persistent storage.
```

Comment:
Do we actually have tests for this? If not, it'd be good to have them.

Reply:
We do have some for the Yul opcode. I don't remember any for the `transient` keyword though.